drivers/rf-transceiver/madura/adrv9025.c

```diff
@@ -0,0 +1,1178 @@
+			.post_state_sysref = true,
+		},
+		[JESD204_OP_LINK_RUNNING] = {
+			.per_link = adrv9025_jesd204_link_running,
+		},
+		[JESD204_OP_OPT_POST_RUNNING_STAGE] = {
+			.per_device = adrv9025_jesd204_post_running_stage,
+			.mode = JESD204_STATE_OP_MODE_PER_DEVICE,
+		},
+	},
+
+	.max_num_links = 5,
+	.sizeof_priv = sizeof(struct adrv9025_jesd204_priv),
+};
+
+/**
+ * Initialize the device.
+ * @param dev - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adrv9025_init(struct adrv9025_rf_phy **dev,
+		      const struct adrv9025_init_param *init_param)
+{
+	adi_adrv9025_SpiSettings_t *spi_settings;
+	struct adrv9025_jesd204_priv *priv;
+	struct adrv9025_rf_phy *phy;
+	int ret;
+
+	phy = (struct adrv9025_rf_phy *)no_os_calloc(1, sizeof(*phy));
+	if (!phy)
+		goto error;
+
+	phy->madDevice = init_param->adrv9025_device;
+	phy->spi_device_id = 0;
+	phy->device_id = 0;
+	phy->dev_clk = init_param->dev_clk;
+
+	spi_settings = adrv9025_spi_settings_get();
+	phy->spiSettings.msbFirst = spi_settings->msbFirst;
+	phy->spiSettings.enSpiStreaming = spi_settings->enSpiStreaming;
+	phy->spiSettings.autoIncAddrUp = spi_settings->autoIncAddrUp;
+	phy->spiSettings.fourWireMode = spi_settings->fourWireMode;
+	phy->spiSettings.cmosPadDrvStrength = spi_settings->cmosPadDrvStrength;
+
+	ret = adrv9025_setup(phy);
+	if (ret < 0) {
+		pr_err("%s: adrv9025_setup failed (%d)\n", __func__, ret);
+		goto error_setup;
+	}
+
+	ret = jesd204_dev_register(&phy->jdev, &jesd204_adrv9025_init);
+	if (ret)
+		goto error_setup;
+
+	priv = jesd204_dev_priv(phy->jdev);
+	priv->phy = phy;
+
+	*dev = phy;
+
+	return 0;
+error_setup:
+	no_os_free(phy);
+	phy = NULL;
+error:
+	return ret;
+}
+
+static int32_t adrv9025_bb_recalc_rate(struct no_os_clk_desc *desc,
+				       uint64_t *rate)
+{
+	struct adrv9025_rf_phy *adrv9025_dev;
+	uint64_t read_rate = 0;
+	int ret;
+
+	adrv9025_dev = desc->dev_desc;
+
+	ret = no_os_clk_recalc_rate(adrv9025_dev->dev_clk, &read_rate);
+	if (!ret)
+		*rate = read_rate;
+
+	return ret;
+}
+
+static int32_t adrv9025_bb_set_rate(struct no_os_clk_desc *desc,
+				    uint64_t rate)
+{
+	struct adrv9025_rf_phy *adrv9025_dev;
+	adrv9025_dev = desc->dev_desc;
+
+	return no_os_clk_set_rate(adrv9025_dev->dev_clk, rate);
+}
+
+static int32_t adrv9025_bb_round_rate(struct no_os_clk_desc *desc,
+				      uint64_t rate,
+				      uint64_t *rounded_rate)
+{
+	pr_debug("%s: Rate %lu Hz", __func__, rate);
+
+	*rounded_rate = rate;
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the CLK structure.
+ *
+ * @param desc - The CLK descriptor.
+ * @param init_param - The structure holding the device initial parameters.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int adrv9025_bb_clk_init(struct no_os_clk_desc **desc,
+				const struct no_os_clk_init_param *init_param)
```

Comment:
Similar comments as for AD9528 (remove clk_init and clk_remove).

Reply:
Done.